src/rust/ide/src/controller/visualization.rs

```diff
@@ -0,0 +1,254 @@
+        let data = Rc::new(RefCell::new(data));
+        Self {data}
+    }
+
+    async fn list_file_visualizations(&self) -> FallibleResult<Vec<VisualizationIdentifier>> {
+        let data                = self.data.borrow();
+        let language_server_rpc = &data.language_server_rpc;
+        let root_id             = language_server_rpc.content_root();
+        let path                = language_server::Path::new(root_id,&[VISUALISATION_FOLDER]);
+        let file_list           = language_server_rpc.file_list(&path).await?;
+        let result              = file_list.paths.iter().filter_map(|object| {
+            if let language_server::FileSystemObject::File{name,path} = object {
+                let mut path = path.clone();
+                path.segments.push(name.to_string());
+                let root_id    = path.root_id;
+                let segments   = path.segments;
+                let path       = language_server::Path{root_id,segments};
+                let identifier = VisualizationIdentifier::File(path);
+                Some(identifier)
+            } else {
+                None
+            }
+        }).collect();
+        Ok(result)
+    }
+
+    fn list_embedded_visualizations(&self) -> Vec<VisualizationIdentifier> {
+        let data   = self.data.borrow();
+        let result = data.embedded_visualizations.keys().cloned();
+        let result = result.map(VisualizationIdentifier::Embedded);
+        result.collect()
+    }
+
+    /// Get a list of all available visualizations.
+    pub async fn list_visualizations(&self) -> FallibleResult<Vec<VisualizationIdentifier>> {
+        let mut visualizations = self.list_embedded_visualizations();
+        visualizations.extend_from_slice(&self.list_file_visualizations().await?);
+        Ok(visualizations)
+    }
+
+    /// Load the source code of the specified visualization.
+    pub async fn load_visualization
+    (&self, visualization:&VisualizationIdentifier) -> FallibleResult<Rc<class::Handle>> {
+        match visualization {
+            VisualizationIdentifier::Embedded(identifier) => {
+                let data       = self.data.borrow();
+                let result     = data.embedded_visualizations.get(identifier);
+                let identifier = visualization.clone();
+                let error      = || VisualizationError::NotFound{identifier}.into();
+                result.cloned().ok_or_else(error)
+            },
+            VisualizationIdentifier::File(path) => {
+                let language_server = &self.data.borrow().language_server_rpc;
+                let js_code         = language_server.read_file(&path).await?.contents;
+                let identifier      = visualization.clone();
+                let error           = |_| VisualizationError::CompileError{identifier}.into();
+                let js_class        = JsSourceClass::from_js_source_raw(&js_code).map_err(error);
```

Comment:
The `JsSourceClass` seems to refer to this as "Instantiation Error" rather than "Compile Error"